packages/upload/src/vaadin-upload-file.js

```diff
@@ -143,6 +158,38 @@ class UploadFile extends ThemableMixin(PolymerElement) {
     ];
   }
 
+  /** @protected */
+  ready() {
+    super.ready();
+
+    // Handle moving focus to the button on Tab.
+    this.shadowRoot.addEventListener('focusin', (e) => {
+      const target = e.composedPath()[0];
+
+      if (target.getAttribute('part').endsWith('button')) {
+        this._setFocused(false);
+      }
+    });
+
+    // Handle moving focus from the button on Shift Tab.
+    this.shadowRoot.addEventListener('focusout', (e) => {
+      if (e.relatedTarget === this) {
+        this._setFocused(true);
+      }
+    });
+  }
+
+  /**
+   * Override method inherited from `FocusMixin` to mark the file as focused
+   * only when the host is focused.
+   * @param {Event} event
+   * @return {boolean}
+   * @protected
+   */
+  _shouldSetFocus(event) {
+    return event.composedPath()[0] === this;
```

Comment:
This doesn't seem to be tested. Quickly tried with the following:

```js
it('should not add focus-ring when a button is shift-tabbed to', async () => {
  // Tab out of the file element
  await sendKeys({ press: 'Tab' });
  await sendKeys({ press: 'Tab' });
  await sendKeys({ press: 'Tab' });

  // Shift-tab back to the file element button
  await sendKeys({ down: 'Shift' });
  await sendKeys({ press: 'Tab' });
  await sendKeys({ up: 'Shift' });

  expect(fileElement.hasAttribute('focus-ring')).to.be.false;
});
```

Reply:
Good point, thanks. I will add the missing test.